### ISSUE TITLE
[PM] Audio device failed to suspend with HDA support enabled on APL-nocodec

### ISSUE THREAD

**Issue by markyang** [thesofproject/linux#512]
**Summary:**
  Check runtime_status, it always shows "active" whether audio is working or not on APL-nocodec.
   It's okay on APL PCM512x.

**Test steps:**
```
1. cat /sys/bus/pci/devices/0000:00:0e.0/power/runtime_status
active
```

**Test env:**
sof master: c6ff3fd5
kernel sof-dev: 5931a304
tplg: sof-apl-nocodec.tplg

**Comment by bardliao** [comment]
@markyang I believe that the issue can be fixed by adding HDMI support to nocodec machine driver and topology. The question is do we really need to enable HDMI support in the kconfig if we don't use it?

**Comment by mengdonglin** [comment]
@bardliao This can be a solution. But probably we need a clean isolation of HDA-stuff in case we want to use the nocodec machine driver/topology to verify non-HDA audio on soe platforms that don't require HDA. In addition, HDMI audio depends on i915 display driver and usually i915 is not stable for early platforms. We don't want audio to be impacted.

@plbossart Would you please comment?

**Comment by plbossart** [comment]
On 1/8/19 8:29 PM, Mengdong Lin wrote:
>
> @bardliao <https://github.com/bardliao> This can be a solution. But 
> probably we need a clean isolation of HDA-stuff in case we want to use 
> the nocodec machine driver/topology to verify non-HDA audio on soe 
> platforms that don't require HDA. In addition, HDMI audio depends on 
> i915 display driver and usually i915 is not stable for early 
> platforms. We don't want audio to be impacted.
>
> @plbossart <https://github.com/plbossart> Would you please comment?
>
There's already a Kconfig to take HDaudio links out (which removes HDMI) 
and another to take out HDaudio codecs. Not sure what else you need for 
testing?

If the ask is to add HDMI to the nocodec stuff by default, then we'd 
need a longer discussion. I'd prefer to handle HDMI as a separate card 
that can be added without having to copy/paste all the time. Or at least 
we need HDMI to be more like a turn-key component that can be added in a 
simpler way than mucking with the machine driver.

**Comment by mengdonglin** [comment]
@xiulipan @xun2z Could you help to check if current kconfig used by daily build server has disabled both HDaudio links and HDaudio codecs (both legacy HD-A codecs and asoc HDMI audio codec) ?

**Comment by xun2z** [comment]
@mengdonglin Kconfig is OK. nocodec is isolated well from HDA.
This issue is caused by no `pm_runtime_suspend` support in nocodec, we have only machine driver. As pierre said, we don't want to add hard-coded HDA. I perfer to try a virtual codec driver supporting PM, or disable PM in nocodec.

**Comment by lyakh** [comment]
@plbossart what would be your preferred way to solve this? According to @mengdonglin 's recommendations - adding a virtual codec or disabling PM for nocodec? How important is PM for nocodec?

**Comment by plbossart** [comment]
@ranj063 simplified the whole suspend scheme now and I don't think we need to do anything special.
Can you retry and confirm the diagnosis?

**Comment by lyakh** [comment]
@plbossart Retried, that hasn't fixed it (unfortunately).

**Comment by ranj063** [comment]
@lyakh this is a bit surprising. Just to confirm, you're using sof-apl-nocodec.tplg and the kernel is latest tip of sof-dev?
Also do you mind sharing your kconfig?

**Comment by lyakh** [comment]
@ranj063 Yes, sof-dev tip - with your 2 PM patches, nocodec topology. I can also attach dmesg. I see there that volume runtime-suspend is failing: `error: volume put failed to idle -16` .config attached
[.config.txt](https://github.com/thesofproject/linux/files/2837107/default.config.txt)

**Comment by lyakh** [comment]
I know why (I think) - the nocodec platform driver is registered as a child of the PCI device and then rt-suspend is rejected, since the .ignore_children flag isn't set. Expect a patch, if that's the only problem...

**Comment by lyakh** [comment]
yep, fixed it by setting .ignore_children, but we have to consider what consequences this can have. I think it should be ok, I don't see the PCI device getting any more children than the nocodec device, but maybe I'm missing some. I'll submit a patch, we can discuss it there then.

**Comment by plbossart** [comment]
That doesn't seem right to me. Yes the nocodec platform driver is registered as a platform driver, but so are *all* the machine drivers. In other words the existing logic works for most boards without this .ignore_children. we must be missing something else.

**Comment by plbossart** [comment]
I just tried and see no problem
````
root@ubilinux4:/sys/devices/pci0000:00/0000:00:0e.0# more power/runtime_status 
suspended
root@ubilinux4:/sys/devices/pci0000:00/0000:00:0e.0# more sof-nocodec/power/runtime_status 
unsupported
````

**Comment by plbossart** [comment]
I can even see the transition from active to suspend, it happens with a delay with is normal with autosuspend.
````

speaker-test 1.1.3

Playback device is hw:0,0
Stream parameters are 48000Hz, S16_LE, 2 channels
Sine wave rate is 440.0000Hz
Rate set to 48000Hz (requested 48000Hz)
Buffer size range from 384 to 16320
Period size range from 192 to 4032
Using max buffer size 16320
Periods = 4
was set period_size = 3456
was set buffer_size = 16320
 0 - Front Left
 1 - Front Right
Time per period = 5.708378
 0 - Front Left
 1 - Front Right
^CTime per period = 4.124723
root@ubilinux4:/sys/devices/pci0000:00/0000:00:0e.0# more power/runtime_status 
active
root@ubilinux4:/sys/devices/pci0000:00/0000:00:0e.0# more power/runtime_status 
suspended
````

**Comment by lyakh** [comment]
@plbossart hm, interesting. And you aren't getting `error: volume put failed to idle -16` in your dmesg? I missed the point, that other machines' platform devices are also children of the PCI device. But I see error messages in my set up and I get -EBUSY from rpm_check_suspend_allowed(). So, it must be the `atomic_read(&dev->power.child_count)` part that's different then. I'll track it down tomorrow.

**Comment by lyakh** [comment]
@plbossart @ranj063 ok, here goes my updated analysis: as @plbossart correctly noted, it isn't about the nocodec vs. pcm5112x topology. It's about whether HDMI codecs are included in the topology or not. When the HDA driver initialises, it checks what codecs are connected by reading some hardware register. Then for each detected codec it allocates objects *and* in snd_hdac_device_init() runtime PM is activated for each of those codecs. Here's a comment from that function:
```
 * This function increments the runtime PM counter and marks it active.
 * The caller needs to turn it off appropriately later.
```
If the topology contains respective codecs then exactly that happens - topology drivers perform their initialisation and then decrement the PM counter. But if codec drivers don't probe, that doesn't happen.
Now, how do we fix it. First of all, this doesn't seem a robust design to me - enabling PM in one driver and letting other drivers disable it. There can be many reasons why those drivers might not complete their probing, which will always lead to the hardware remaining active. Especially with the automatic suspend and with a sufficient delay, the HDA driver could safely do a put() and if the codec driver registers soon enough (which should typically be the case), power will actually stay on. I think that would be the right solution. But it would be rather intrusive - the driver is quite generic.
Otherwise, if there are strong reasons to keep the present design, we have to think how to fix that at least for our nocodec configurations.

**Comment by ranj063** [comment]
@lyakh but there are no HDMI pipelines in the nocodec topology right? The nocodec machine driver cannot handle HDMI codecs. So Im not sure ho you're even seeing this problem

**Comment by lyakh** [comment]
@ranj063 the HDA driver is initialising anyway as a PCI device

**Comment by ranj063** [comment]
@lyakh ahh i get the point now. I am not sure how to handle this but this seems a bit of an overkill to handle such cases. When using nocodec tplg, we should disable HDA in the kernel.

Let me test this at my end today and confirm this problem anyway.

**Comment by lyakh** [comment]
@ranj063 well, sure, we can make sure to avoid this issue *when* using nocodec, but is that enough? I think there are multiple other situations when for one reason or for another a codec driver would not probe correctly, would this be a reason enough to keep the HDA device active? Since the HDA driver isn't SOF specific, I think we can do some preliminary discussing here and then take this to the ALSA ML?

**Comment by plbossart** [comment]
I am having difficulties following this thread. We already have multiple Kconfig options to support HDaudio and nocodec:

SND_SOC_SOF_NOCODEC
SND_SOC_SOF_HDA_LINK
SND_SOC_SOF_HDA_AUDIO_CODEC

We don't look for the codecs in all cases, so what combination of these three exposes the problem on Up2?

**Comment by lyakh** [comment]
@plbossart I think I attached my .config above ;-) But anyway - enable and force nocodec and keep all the HDA options too. That should do it.

**Comment by plbossart** [comment]
@lyakh your initial explanations threw me off ("It's about whether HDMI codecs are included in the topology or not"). This is really not about the topology file but about the Kconfigs. If we have the HDaudio link enabled, then indeed we will initialize codecs but the card will never make any reference to those codecs so we can't suspend them, right?

if we are aligned on the analysis, I see three ways of fixing this.
1. we make NOCODEC incompatible with HDA_LINK - and move the NOCODEC to a different defconfig that's not selected. simple but we lose a bit in terms of flexibility.
2. we are smarter and modify hda_init_caps() so that when only a single codec is detected but we are in NOCODEC mode we don't initialize those codecs.
3. we add another option where we specify that we want NOCODEC to support suspend/resume, and this disables HDA_LINK. That might be the best case for testing.

**Comment by ranj063** [comment]
@plbossart I agree option 3 sounds like the best thing to do.

**Comment by lyakh** [comment]
To choose one of the above options I'd like to understand - is this really only a problem with nocodec? What if nocodec isn't used but the real codec, that is supposed to probe, hasn't succeeded for some reason, or haven't even attempted? Isn't it better to let the HDA driver autosuspend itself at the end of its probing and let any further users (codecs) "get()" it again?

**Comment by bardliao** [comment]
Agree with @lyakh This issue will happen if SND_SOC_HDAC_HDMI is selected but HDMI is not used. Similar issue is discussed in https://github.com/thesofproject/linux/issues/140#issuecomment-430607566

**Comment by xun2z** [comment]
Making CONFIG_SND_SOC_SOF_HDA_LINK incompatible with nocodec in kconfig is a solution for this issue. will close this issue when corresponding patch is merged
